python-api/livy/client.py

```diff
@@ -0,0 +1,167 @@
+
+        Note that the URL should be reachable by the Spark driver process. If running the driver
+        in cluster mode, it may reside on a different host, meaning "file:" URLs have to exist
+        on that node (and not on the client machine).
+
+        :param file_uri: Representation of the path to a file using the filesystem encoding
+        :type file_uri: PurePath from pathlib or pathlib2(preferred)
+
+        >>> import pathlib2
+        >>> pathlib2.Path("/test.txt").asuri()
+
+        :returns A future to monitor the status of the job
+        """
+        pass
+
+    def add_pyfile(self, file_uri):
+        """
+        Adds a .py or .zip to the running remote context.
+
+        Note that the URL should be reachable by the Spark driver process. If running the driver
+        in cluster mode, it may reside on a different host, meaning "file:" URLs have to exist
+        on that node (and not on the client machine).
+
+        :param file_uri: Representation of the path to a file using the filesystem encoding
+        :type file_uri: PurePath from pathlib or pathlib2(preferred)
+
+        >>> import pathlib2
+        >>> pathlib2.Path("/test.txt").asuri()
+
+        :returns A future to monitor the status of the job
+        """
+        pass
+
+    def upload_file(self, open_file):
+        """
+        Upload a file to be passed to the Spark application
+        :param file open_file: The local file to be uploaded
+        :returns A future that can be used to monitor this operation
+        """
+        pass
+
+    def upload_pyfile(self, open_file):
+        """
+        Upload a .py or .zip dependency to be passed to the Spark application
+        :param file open_file: The local file to be uploaded
+        :returns A future that can be used to monitor this operation
+        """
+        pass
+
+    def __set_uri(self, uri):
+        if uri is not None and ('http' == uri.scheme or 'https' == uri.scheme):
+            self.__config.set(self.__CONFIG_SECTION, 'livy.uri', uri.geturl())
+        else:
+            url_exception = uri.geturl if uri is not None else None
+            raise Exception('Cannot create client - Uri not supported - ', url_exception)
+
+    def __set_conf(self, key, value):
+        if value is not None:
+            self.__config.set(self.__CONFIG_SECTION, key, value)
+        else:
+            self.__config.remove_option('env', key)
+
+    def __set_multiple_conf(self, conf_dict):
+        if conf_dict is not None:
+            for key, value in conf_dict.iteritems():
+                self.set_conf(self.__CONFIG_SECTION, key, value)
+
+    def __load_config(self, load_defaults, conf_dict):
+        self.__config.add_section(self.__CONFIG_SECTION)
+        if load_defaults is True:
+            self.__handle_default_config_loading()
+        self.__set_multiple_conf(conf_dict)
+
+    def __handle_default_config_loading(self):
+        has_config_dir = os.environ.has_key(self.__LIVY_CLIENT_CONF_DIR)
```

Comment:
As above, consider making `livy_client_conf_dir` an attribute of the object that's set at creation, rather than looking it up in the environment. All else equal, IMO mixing in config from the environment is confusing. You might need to use the environment for consistency with other Livy clients.